### PR TITLE
precompute document-term matrix

### PR DESCRIPTION
**Pull request by flinder**
@Michael98Liu let's use this pull request to discuss the code you wrote. 

There are a few minor issues that I've pointed out in the code (see the comments). The error you mentioned is caused by using `np.hstack` instead of `scipy.sparse.hstack`.

However, when fixing that, the code breaks due to a more serious issue: 
When we run:
```
            X = dataset.get_texts('train')
            y = dataset.get_labels('train')
            CV.fit(X, y)
```
All training data (i.e. all 'rows' of `X`) are passed to the CV. CV splits `X` and `y` into the training and validation set for each fold. But because we then ignore the `X` (or, you named it `rawdocuments`) in `MyVectorizer.transform` and instead load the cached pre-computed matrix from the dimensions of `X` and `y` won't match anymore. That is, `y` is passed on from the `train_test_split` (i.e. it only has the training data for the current fold but not the validation data), whereas the `X` we load from disk has the complete data. 

Not sure if that's the ideal way, but one solution would be the following (pseudocode, not tested):
```
MyVectorizer()
    def fit_transform(self, X, y=None):
        train_indices = X.index
        X_pre = load_precomputed_matrices() # this does the merging of the ngram matrices
        return X_pre[train_indices, :]
```
Lmk what you think.

### CODE REVIEW

File: paper_experiment/experiment.py [hunk 1]
```diff
@@ -64,36 +66,38 @@
...
 
     models = ['svm', 'lasso', 'random_forest']
     for dataset, model in itertools.product(data_sets, models):
+        transform_and_cache(dataset)
+
         for feature_set in ['word_ngrams', 'char_ngrams', 'embeddings']:
             print(f"Training {model} on {dataset.name} with {feature_set}")
-            M = Models(model, feature_set, dataset.tokenizer, MAX_N_FEATURES)
+            M = Models(dataset, model, feature_set, dataset.tokenizer, MAX_N_FEATURES)
             n_iter = 20
```
Comment (flinder) [inline paper_experiment/experiment.py]: This line is redundant

File: paper_experiment/models.py [hunk 2]
```diff
@@ -173,17 +174,41 @@ def fit_transform(self, X, y=None, **fit_params):
...
+        self.ngram_range = ngram_range
+
+    def fit_transform(self, rawdocuments, y):
+        if self.feature_set in ['char_ngrams', 'word_ngrams']:
+            # load and concatenate document-term matrices
+            #X = pickle.load(open("./cache/{}_{}_{}".format(self.dataset.name, self.analyzer, self.ngram_range[0], "rb") ))
+            X = sparse.load_npz("./cache/{}_{}_{}.npz".format(self.dataset.name, self.analyzer, self.ngram_range[0]))
+            for i in range(self.ngram_range[0]+1, self.ngram_range[1]+1):
+                #X1 = pickle.load(open("./cache/{}_{}_{}".format(self.dataset.name, self.dataset.analyzer, i, "rb") ))
+                X1 = sparse.load_npz("./cache/{}_{}_{}.npz".format(self.dataset.name, self.analyzer, self.ngram_range[0]))
```
Comment (flinder) [inline paper_experiment/models.py]: `self.ngram_range[0]` should be just `i` (if not it loads the same matrix in every iteration)

File: paper_experiment/models.py [hunk 3]
```diff
@@ -173,17 +174,41 @@ def fit_transform(self, X, y=None, **fit_params):
...
+        elif feature_set == 'word_ngrams':
+            self.analyzer = 'word'
+        self.feature_set = feature_set
+        self.ngram_range = ngram_range
+
+    def fit_transform(self, rawdocuments, y):
+        if self.feature_set in ['char_ngrams', 'word_ngrams']:
+            # load and concatenate document-term matrices
+            #X = pickle.load(open("./cache/{}_{}_{}".format(self.dataset.name, self.analyzer, self.ngram_range[0], "rb") ))
+            X = sparse.load_npz("./cache/{}_{}_{}.npz".format(self.dataset.name, self.analyzer, self.ngram_range[0]))
```
Comment (flinder) [inline paper_experiment/models.py]: Minor comment: Can you make sure to have maximum line length of 80 characters (there will be more explicit code style guidelines in the lab soon). 
Here a f-string would make it a bit shorter:
`X = sparse.load_npz(f'./cache/{self.dataset.name}_{self.analyzer}_{self.n_gram_range[0]}')`

File: paper_experiment/models.py [hunk 4]
```diff
@@ -173,17 +174,41 @@ def fit_transform(self, X, y=None, **fit_params):
...
+            return X
 
 class Models:
-    def __init__(self, name, feature_set, tokenizer, max_n_features):
+    def __init__(self, dataset, name, feature_set, tokenizer, max_n_features):
         self.name = name
+        self.dataset = dataset
         self.max_n_features = max_n_features
         self.tokenizer = tokenizer
         if feature_set in ['char_ngrams', 'word_ngrams']:
```
Comment (flinder) [inline paper_experiment/models.py]: This is an enhancement but probably good to do right away: Here it will make sense to split up the param ranges for `'char_ngrams'` and `'word_ngram'` since we most like won't want the same ranges for both feature types (i.e. character 1-grams don't seem very informative to me also 5-grams for words seems excessive). Before we had for word ngrams: 1 to 3 and for character ngrams 3 to 5

File: paper_experiment/models.py [hunk 5]
```diff
@@ -206,7 +231,7 @@ def __init__(self, name, feature_set, tokenizer, max_n_features):
             self.classifier = GaussianNB()
             self.params.update({'clf__var_smoothing': sp.stats.uniform(1e-12, 1e-6)})
         elif self.name == "svm":
-            self.classifier = SVC()
+            self.classifier = SVC(gamma = "scale")
             self.params.update({'clf__kernel': ['linear', 'rbf'],
```
Comment (flinder) [inline paper_experiment/models.py]: I think for SVM the following parameter grid would make more sense: 
```
self.classifier = SVC(kernel='rbf') 
self.params.update({'clf__gamma': sp.stats.uniform(1e-3, 1e3),
                                   'clf__C': sp.stats.uniform(1e-2, 1e2)})
```
Reasons:
- `'rbf'` supersedes the linear kernel (i.e. the linear kernel is a specific type of rbf kernel because the non-parametric kernel can 'mimic' a linear kernel. see here: https://stats.stackexchange.com/questions/73032/linear-kernel-and-non-linear-kernel-for-support-vector-machine)
- gamma and C are actually the more interesting tuning parameters (see https://scikit-learn.org/stable/auto_examples/svm/plot_rbf_parameters.html)
- you can keep the `tol` grid but it's probably not hugely important, so we could also set it to avoid warnings

File: paper_experiment/models.py [hunk 6]
```diff
@@ -236,3 +259,14 @@ def __init__(self, name, feature_set, tokenizer, max_n_features):
             raise ValueError(
                 ("feature_set must be one of ['char_ngrams', 'word_ngrams',"
                  "'embeddings]"))
+
+def transform_and_cache(dataset):
```
Comment (flinder) [inline paper_experiment/models.py]: - This function needs a docstring

File: paper_experiment/models.py (hunk 3 above)
Comment (flinder) [reply paper_experiment/models.py]: You can run (from the terminal): `pylint filename.py` to get some feedback on coding style (you don't have to fix everything it reports in default mode (e.g. constant names) but it gives you an idea of issues)

File: paper_experiment/models.py [hunk 7]
```diff
@@ -173,17 +174,41 @@ def fit_transform(self, X, y=None, **fit_params):
     def fit(self, X, y=None, **fit_params):
         return self
 
+class MyVectorizer(CountVectorizer):
```
Comment (flinder) [inline paper_experiment/models.py]: Maybe change this name to something like `PrecomputeVectorizer` . Just something more expressive than this generic name

Comment (flinder) [reply paper_experiment/models.py]: Also a docstring with minimal documentation would be good
